### ISSUE TITLE
The new code generator adds everything to feature.AddKnownType

### ISSUE THREAD

**Issue by pentp** [dotnet/orleans#5021]
This was a problem with the old code generator already as documented in #4437, but with the new code generator it's a lot worse.

For example I have a functional tests project that has a single grain interface with a few simple calls, it references various other projects. Both code generators write about 200 lines of code for the grain method invoker, but then they generate endless `feature.AddKnownType` lines - the old one about 700 lines, the new one about **26000** lines! The generated assembly grows from 300KB to 6MB and is really slow to start up.

What's the purpose of `feature.AddKnownType`? It looks like it adds every type that is visible to Roslyn during codegen or something similar. And the added information doesn't look useful at all - it's just a list of type names.

**Comment by ReubenBond** [comment]
Thanks for reporting, @pentp. I've opened #5026 with a fix. Hope it works well for you. There will still be some (maybe many) `AddKnownType` lines for assemblies which are referenced by your project but not by other projects (eg not referenced by Orleans), but the number should be reduced so that there are less duplicates.

This is a side effect of how type serialization works which I hope to remove in a future release, maybe 3.0.

**Comment by ReubenBond** [comment]
@pentp this is a side effect of how type serialization works - it needs to know the list of all possible types before it encounters them during deserialization. If there's a *possibility* that a type might be serialized, it's included. That means that very few types are excluded.

We could consider a switch which tells the code generator to modify its behavior and try to be more conservative about which types it includes.

Ultimately, it would be good to entirely avoid this process by using a different method for type serialization.

Scanning all possible types at runtime would force all assemblies and transitive dependencies to be loaded during startup.

Alternatively, what do you suggest?

**Comment by ReubenBond** [comment]
@pentp this is a side effect of how type serialization works - it needs to know the list of all possible types before it encounters them during deserialization. If there's a *possibility* that a type might be serialized, it's included. That means that very few types are excluded.

We could consider a switch which tells the code generator to modify its behavior and try to be more conservative about which types it includes.

Ultimately, it would be good to entirely avoid this process by using a different method for type serialization.

Scanning all possible types at runtime would force all assemblies and transitive dependencies to be loaded during startup (which is slow).

I'm open to suggestions.

**Comment by pentp** [comment]
The old code generator included a lot less types, so it was tolerable (and still worked) - maybe the list of possible types can be reasonably limited?  I have some stats for known types of old vs new:

Project | Old | New
------------ | ------------ | -------------
Interfaces | 5053 | 7876
Actors | 4295 | 25554
FunctionalTests | 721 | 1208

I propose some improvements:

- Excluding some (most) types from the list. Almost all types there are obviously not needed, but unfortunately there is no obvious way to filter them out... At least interfaces, static classes and abstract classes could be excluded.
- Provide a way for users to opt-out of some (most) of the types, e.g. include only public types, include only types from a fixed set of assemblies.
- Exclude `IFeaturePopulator` generated code from PDB by using `#line hidden` - unlike serialization/invocation code, it doesn't need debugger support.
- A compact encoding for smaller DLL size and faster startup, e.g. types grouped by assembly, sorted by namespace+name with shared prefix, binary encoding (byte[] array), etc.
- Change the wire format to include the assembly name, with encoding formats something like this:
    - `NamedTypeKnownAssembly`: A predefined list of System/Orleans/common assemblies with fixed IDs.
    - `NamedTypeWithAssemblyPrefix`: Assembly name deduced from type name (very common) - only number of "dots" needs to be encoded.
    - `NamedType`: Fully qualified type name encoded.

**Comment by ReubenBond** [comment]
> At least interfaces, static classes and abstract classes could be excluded.

Not true for interfaces or abstract classes. Even though a type might never appear on the wire, it can appear as the generic parameter of another type. For example, you might need to serialize `List<IMyInterface>` or `List<MyAbstractClass>`. I'll add a filter for static types - thanks for pointing that out.

> Alternatively, change the wire format [...]

Wire format changes won't affect this until we deprecate the existing format and make it optional. I agree with you, changes like that could be beneficial. I'd like to overhaul type encoding and (serialization in general) for 3.0. I've been experimenting with changes similar to the suggested `NamedTypeIndex` encoding in a side project and hope to wind some/all of that into Orleans.

> Provide a way for users to opt-out of some (most) of the types, e.g. include only public types, include only types from a fixed set of assemblies.

This is the option I'm currently most amenable to. Would a csproj property suffice? Eg:

``` xml
<PropertyGroup>
 <OrleansCodeGenIgnoreAssemblies>
    $(OrleansCodeGenIgnoreAssemblies);System.Fabric
  </<OrleansCodeGenIgnoreAssemblies>
</PropertyGroup>
```

The property could be an exact match on assembly name. How would you prefer it be exposed?

**Comment by pentp** [comment]
I'm not sure if whitelisting or blacklisting is better here, but blacklisting on csproj level should work, provided that all the referenced assemblies are ignored also (unless some non-ignored assembly references them). That probably needs to be encoded in a assembly attribute also, so that if I ignore assembly X in project A and then run Orleans code generator for project B (which references A) it doesn't generate anything for assembly X again.

In short, if I use assembly X in a project and add `OrleansCodeGenIgnoreAssemblies` for it then I shouldn't have to worry about ignoring it or any of its dependencies again (the list of dependencies could easily change if it comes from nuget).

**Comment by ReubenBond** [comment]
Fixed by #5036

**Comment by pentp** [comment]
#5036 doesn't have any practical effect on this issue, it's from 2018, did you mean some new PR instead?
The new code generator still includes tens of thousand of lines of useless crap, so I'm forced to use the old generator that at least generates only about 10000 of these...

To actually fix this the wire format needs to be extended a bit to fully qualify the type names (include assembly name) and at the same time type name de-duplication could be done also.

**Comment by ReubenBond** [comment]
That's fair, ideally we can remove them altogether over time.

> To actually fix this the wire format needs to be extended a bit to fully qualify the type names (include assembly name) and at the same time type name de-duplication could be done also.

I added some experimental support for that for testing purposes but it needs more work to be sound in all cases. Now that 3.0 is wrapping up I hope we can invest more in this area
